### PR TITLE
Add Cuda support

### PR DESCRIPTION
**Pull request by jpakkane**
This is #3730 rebased, fixed and with better tests.

I'm not sure how we should do the testing on this. The legalese on NVidia's web site seems to indicate that they don't like redistributing it and the package itself does not even install on Ubuntu bionic. There are Cuda packages in universe that work, but they add a gigabyte or so of stuff making the CI image even more bloated than it is currently.

### CODE REVIEW

File: mesonbuild/compilers/__init__.py [hunk 1]
```diff
@@ -154,6 +155,7 @@
     GnuDCompiler,
     LLVMDCompiler,
 )
+from .cuda import CudaCompiler
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/__init__.py]: __[Flake8]__

[F401] '.cuda.CudaCompiler' imported but unused

File: mesonbuild/compilers/compilers.py [hunk 2]
```diff
@@ -137,6 +139,13 @@ def is_library(fname):
                            'release': ['-Os'],
                            'minsize': ['-Oz']}
 
+cuda_buildtype_args = {'plain': [],
+                       'debug': ['-O0', '-g'],
+                       'debugoptimized': ['-O1', '--debug'],
+                       'release': ['-O3', '-Otime'],
+                       'minsize': ['-O3', '-Ospace'],
+                      }
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/compilers.py]: __[Flake8]__

[E124] closing bracket does not match visual indentation

**Comment by jpakkane** [comment]
This does not support dependencies properly because NVidia's compiler does not produce .d files. Instead it requires that you run the compiler again with different command line options to get the dep files. Ninja does not support that and it is wasteful nonetheless. IIRC there are people inside NVidia trying to get this fixed but the status of that is unknown.

**Comment by nirbheek** [comment]
> There are Cuda packages in universe that work, but they add a gigabyte or so of stuff making the CI image even more bloated than it is currently.

We could spin up another ubuntu image for just this and add it as a CircleCI job (so it won't even increase our CI times).

File: mesonbuild/compilers/compilers.py [hunk 3]
```diff
@@ -143,6 +145,13 @@ def is_library(fname):
                            'custom': [],
                            }
 
+cuda_buildtype_args = {'plain': [],
+                       'debug': [],
+                       'debugoptimized': [],
+                       'release': [],
+                       'minsize': [],
+                      }
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/compilers.py]: __[Flake8]__ [E124] closing bracket does not match visual indentation

<sub>posted by Sider</sub>

File: mesonbuild/compilers/__init__.py [hunk 4]
```diff
@@ -153,6 +154,7 @@
...
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/__init__.py]: __[Flake8]__ [F401] '.cuda.CudaCompiler' imported but unused

<sub>posted by Sider</sub>

**Pull request merged**